`resolve.tsconfig` path pattern `"@*"` incorrectly matches scoped npm packages

### Have you used AI?

Yes

### Bug Description

When `resolve.tsconfig` is enabled and `tsconfig.json` contains a path mapping such as:

```json
"paths": {
  "@*": ["./src/*"]
}
```

webpack resolves imports like `@sentry/react` as if they were project aliases and tries to load them from `./src/...` under the mapping, instead of resolving them as scoped packages from `node_modules`. The build fails with “Can’t resolve '@sentry/react'” from application directories.

`@sentry/react` is a real scoped npm package, not an app alias. For example **`tsconfig-paths-webpack-plugin`** only rewrites when `tsconfig-paths` finds a match; otherwise resolution continues and `node_modules` still works. **`resolve.tsconfig`** here applies the `@*` mapping first that `@sentry/react` is sent to `./src/*` instead of resolving as a dependency.

### Link to Minimal Reproduction and step to  reproduce

1. Public repo: https://github.com/whoisYeshua/webpack-simple-boilerplate/tree/webpack-typescript
2. `npm i`
3. Run `npm start`.

### Expected Behavior

`@sentry/react` resolves to `node_modules/@sentry/react` like any other scoped dependency

### Actual Behavior

Webpack cannot resolve `@sentry/react` from the app folders below; that matches the import being handled as a `./src/*` alias instead of a package under node_modules::

```bash
ERROR in ./src/hooks/useFetch.ts 6:0-49
Module not found: Error: Can't resolve '@sentry/react' in '.../src/hooks'

ERROR in ./src/index.tsx 8:0-50
Module not found: Error: Can't resolve '@sentry/react' in '.../src'

ERROR in ./src/sentry.ts 4:0-37
Module not found: Error: Can't resolve '@sentry/react' in '.../src'
```

### Environment

```shell
System:
    OS: macOS 15.7.4
    CPU: (10) arm64 Apple M1 Pro
    Memory: 1.07 GB / 16.00 GB
  Binaries:
    Node: 20.19.2 - /Users/konstantin/.nvm/versions/node/v20.19.2/bin/node
    npm: 10.9.4 - /Users/konstantin/Documents/Proga/webpack-simple-boilerplate/node_modules/.bin/npm
  Browsers:
    Chrome: 148.0.7778.96
    Chrome Canary: 150.0.7834.0
    Firefox: 149.0
    Safari: 26.3
    Safari Technology Preview: 26.0
  Packages:
    babel-loader: 10.1.1 => 10.1.1 
    copy-webpack-plugin: 14.0.0 => 14.0.0 
    css-loader: 7.1.4 => 7.1.4 
    css-minimizer-webpack-plugin: 8.0.0 => 8.0.0 
    eslint-webpack-plugin: 6.0.0 => 6.0.0 
    fork-ts-checker-webpack-plugin: 9.1.0 => 9.1.0 
    html-webpack-plugin: 5.6.7 => 5.6.7 
    postcss-loader: 8.2.1 => 8.2.1 
    sass-loader: 16.0.8 => 16.0.8 
    style-loader: 4.0.0 => 4.0.0 
    tsconfig-paths-webpack-plugin: 4.2.0 => 4.2.0 
    webpack: 5.106.2 => 5.106.2 
    webpack-cli: 7.0.2 => 7.0.2 
    webpack-dev-server: 5.2.3 => 5.2.3 
    webpack-merge: 6.0.1 => 6.0.1
```

### Is this a regression?

No

### Last Working Version

_No response_

### Additional Context

_No response_

--- a/.changeset/tsconfig-paths-scoped-pkg-fallthrough.md
+++ b/.changeset/tsconfig-paths-scoped-pkg-fallthrough.md
@@ -1,0 +1,5 @@
+---
+"enhanced-resolve": patch
+---
+
+TsconfigPathsPlugin now falls through to normal module resolution when a `paths` pattern matches but the mapped path does not exist, matching TypeScript's native resolution behavior. Previously, patterns like `"@*"` would block scoped npm packages (e.g. `@sentry/react`) from resolving via `node_modules`.

--- a/lib/TsconfigPathsPlugin.js
+++ b/lib/TsconfigPathsPlugin.js
@@ -272,7 +272,15 @@ module.exports = class TsconfigPathsPlugin {
 							aliasTarget,
 							request,
 							resolveContext,
-							callback,
+							(err, result) => {
+								if (err) return callback(err);
+								if (result) return callback(null, result);
+								// https://github.com/webpack/webpack/issues/20944
+								// Unlike resolve.alias, tsconfig paths should fall through
+								// when a pattern matches but the mapped path does not exist
+								// (matching TypeScript's native resolution behavior).
+								return callback();
+							},
 						);
 					},
 				);

--- a/test/fixtures/tsconfig-paths/scoped-pkg-fallthrough/src/helper/index.ts
+++ b/test/fixtures/tsconfig-paths/scoped-pkg-fallthrough/src/helper/index.ts
@@ -1,0 +1,4 @@
+import * as sentry from "@sentry/react";
+
+export const helper = true;
+export { sentry };

--- a/test/fixtures/tsconfig-paths/scoped-pkg-fallthrough/tsconfig.json
+++ b/test/fixtures/tsconfig-paths/scoped-pkg-fallthrough/tsconfig.json
@@ -1,0 +1,8 @@
+{
+  "compilerOptions": {
+    "baseUrl": ".",
+    "paths": {
+      "@*": ["./src/*"]
+    }
+  }
+}

--- a/test/tsconfig-paths.test.js
+++ b/test/tsconfig-paths.test.js
@@ -1390,6 +1390,59 @@ describe("TsconfigPathsPlugin", () => {
 		});
 	});
 
+	describe("bug: '@*' pattern should fall through for scoped npm packages (#20944)", () => {
+		const scopedPkgDir = path.resolve(
+			__dirname,
+			"fixtures",
+			"tsconfig-paths",
+			"scoped-pkg-fallthrough",
+		);
+
+		it("should resolve '@helper' via the '@*' mapping when the mapped path exists", (done) => {
+			const resolver = ResolverFactory.createResolver({
+				fileSystem,
+				extensions: [".ts", ".tsx", ".js"],
+				mainFields: ["main"],
+				mainFiles: ["index"],
+				tsconfig: path.join(scopedPkgDir, "tsconfig.json"),
+			});
+
+			resolver.resolve({}, scopedPkgDir, "@helper", {}, (err, result) => {
+				if (err) return done(err);
+				if (!result) return done(new Error("No result"));
+				expect(result).toEqual(
+					path.join(scopedPkgDir, "src", "helper", "index.ts"),
+				);
+				done();
+			});
+		});
+
+		it("should fall through to node_modules for '@sentry/react' when '@*' mapping does not resolve", (done) => {
+			const resolver = ResolverFactory.createResolver({
+				fileSystem,
+				extensions: [".ts", ".tsx", ".js"],
+				mainFields: ["main"],
+				mainFiles: ["index"],
+				tsconfig: path.join(scopedPkgDir, "tsconfig.json"),
+			});
+
+			resolver.resolve({}, scopedPkgDir, "@sentry/react", {}, (err, result) => {
+				if (err) return done(err);
+				if (!result) return done(new Error("No result"));
+				expect(result).toEqual(
+					path.join(
+						scopedPkgDir,
+						"node_modules",
+						"@sentry",
+						"react",
+						"index.js",
+					),
+				);
+				done();
+			});
+		});
+	});
+
 	describe("bug: unscoped npm package in extends field", () => {
 		it("should resolve paths inherited from an unscoped npm package tsconfig (extends 'my-base-config')", (done) => {
 			const resolver = ResolverFactory.createResolver({

